### PR TITLE
Support specifying a subset of rows in `GT.data_color()`

### PR DESCRIPTION
**Pull request by jrycw**
Related to #248.

This PR aims to support specifying a subset of rows in `GT.data_color()`. Based on the discussions, it appears that the most common use case involves excluding the last row, which is typically the total row. Consequently, rows that are not selected should not be included in the color "calculation." In this PR, I first select only the specified columns and rows, then "re-map" the "calculated color" back to the original table.

A simple example might illustrate the concept well:
```python
import polars as pl
import polars.selectors as cs
from great_tables import GT

df = pl.DataFrame({"a": [1, 2, 3, 4, 5, 200], "b": [51, 52, 53, 54, 55, 200]})

GT(df).data_color(columns=cs.numeric(), rows=pl.col("a").lt(6), palette="PuBu")
```
image

```python
GT(df).data_color(columns=cs.numeric(), rows=pl.col("a").lt(300), palette="PuBu")
```
image

### Bonus
In addition, I have added support for using Polars-style column and row selection. This allows for more flexible operations, such as:
```python
GT(df).data_color(columns=cs.starts_with("b"), rows=pl.col("a").lt(6), palette="PuBu")
```
image

### CODE REVIEW

File: great_tables/_data_color/base.py [hunk 1]
```diff
@@ -258,21 +266,25 @@ def data_color(
...
-
-            else:
-                gt_obj = gt_obj.tab_style(
-                    style=fill(color=color_val), locations=body(columns=col, rows=[i])
-                )
+        # For each row, we check if the row index is selected, and then apply a fill to the
+        # corresponding cell using `tab_style()`
+        n_rows = len(resolve_rows_i(self))
+        iter_color_vals = iter(color_vals)
+        for i in range(n_rows):
```
Comment (machow) [inline great_tables/_data_color/base.py:273]: Would something like...

```python
for i, color_val in zip(row_pos, color_vals):
    ...
```

work? (in this case, you wouldn't need to check `if i in row_pos`)

**Review by machow — CHANGES_REQUESTED** [review]
Thanks, this looks really great! I left one small question on whether we might be able to tweak the for loop a little bit, but it's very possible I overlooked something there.

File: great_tables/_data_color/base.py (hunk 1 above)
Comment (jrycw) [reply great_tables/_data_color/base.py]: @machow Clever suggestion! I believe this code will work better.

**Comment by jrycw** [comment]
Please hold off on merging this PR for now. I've discovered some personal notes in my Jupyter playbook explaining why I wrote the code this way. It appears that the logic between these two implementations may differ. I'll conduct further investigation soon.

**Comment by jrycw** [comment]
I think both approaches should yield the same result. However, a more interesting question arises: what should users expect if we include the `rows` parameter? For instance, if `rows = [1, 3, 4]`, should we first calculate the palette for the entire column and then retrieve the corresponding colors for those rows, or should we calculate the palette based solely on the selected rows? Both cases might not look aesthetically pleasing, and I doubt their practical application. 

Perhaps what we really need is an option like `include_last_row=True` or `include_last_row=False` (the parameter name might not be optimal, but you get the idea). I've implemented alternative logic in the last commit, although this logic might need adjustments in the future if we introduce something like `GT.summary_rows()`, which could potentially affect more than one row.

```python
import polars as pl
import polars.selectors as cs
from great_tables import GT

df = pl.DataFrame(
    {
        "a": [1, 2, 3, 4, 5, sum([1, 2, 3, 4, 5])],
        "b": [51, 52, 53, 54, 55, sum([51, 52, 53, 54, 55])],
    }
)


GT(df).data_color(columns=cs.numeric(), include_last_row=True, palette="PuBu")
```

image

```python
GT(df).data_color(columns=cs.numeric(), include_last_row=False, palette="PuBu")
```

image

**Comment by machow** [comment]
> should we first calculate the palette for the entire column and then retrieve the corresponding colors for those rows, or should we calculate the palette based solely on the selected rows?

@rich-iannone do you know what happens in the gt R package? AFAICT data_color there doesn't use excluded rows, and the rationale is that this let'd you exclude extremely large values etc..

I would lean toward not using excluded in the calculations. Then, if people want to use all rows, they can always set the `domain=` argument, to specify the exact range for the color values.

**Comment by rich-iannone** [comment]
> > should we first calculate the palette for the entire column and then retrieve the corresponding colors for those rows, or should we calculate the palette based solely on the selected rows?
> 
> @rich-iannone do you know what happens in the gt R package? AFAICT data_color there doesn't use excluded rows, and the rationale is that this let'd you exclude extremely large values etc..
> 
> I would lean toward not using excluded in the calculations. Then, if people want to use all rows, they can always set the `domain=` argument, to specify the exact range for the color values.

The R package doesn't use the excluded rows in the calculations or colorization. You're right that the `domain` argument approach is better for specifying the exact range. If we did include all the rows in the calculation, it might be surprising to users and I don't think it would be generally desired behavior through an extra argument.

**Comment by jrycw** [comment]
Thanks @machow and @rich-iannone . Would [43b0ce5](https://github.com/posit-dev/great-tables/pull/364/commits/43b0ce5c6e5edc1280bcf13b6be8e37a615fc37f) achieve our goal, or are there any other modifications I should make?

**Comment by rich-iannone** [comment]
I think so. Just reverting the last commit would make this right.

**Review by machow — APPROVED** [review]
LGTM, thanks for thinking through the different options here!

**Pull request merged**